File: ios/Gorilla Groove/sync/ServerSynchronizer.swift
import Foundation

typealias PageCompleteCallback = (_ completedPage: Int, _ totalPages: Int, _ type: SyncType) -> Void

class ServerSynchronizer {
    
    static let baseUrl = "sync/entity-type/%@/minimum/%ld/maximum/%ld?size=400&page="
   
    @SettingsBundleStorage(key: "offline_mode_enabled")
    private static var offlineModeEnabled: Bool
    
    private static var syncRunning = false
    
    static var lastSync: Date = {
        var syncStatuses = SyncStatusDao.getStatusByTypes(SyncType.allCases)

        return syncStatuses.map { $0.value.lastSyncAttempted }.max()!
    }()
    
    @discardableResult
    static func syncWithServer(pageCompleteCallback: PageCompleteCallback? = nil, abortIfRecentlySynced: Bool = false) -> Bool {
        if Thread.isMainThread {
            fatalError("Do not sync on the main thread!")
        }
        
        GGSyncLog.info("Initiating sync with server...")
        if syncRunning {
            GGSyncLog.info("Sync already running. Not syncing")
            return false
        }
        
        if offlineModeEnabled {
            GGSyncLog.debug("Not syncing as we are in offline mode")
            return false
        }
                
        if abortIfRecentlySynced {
            if let diff = Calendar.current.dateComponents([.minute], from: lastSync, to: Date()).minute, diff > 10 {
                GGSyncLog.debug("Last sync was old enough to do a new sync")
            }
            else {
                GGSyncLog.debug("Last sync was too recent. Not syncing")
                return false
            }
        }

        syncRunning = true
        
        let newDate = Date()
        let maximum = newDate.toEpochTime()
       
        var syncStatuses = SyncStatusDao.getStatusByTypes(SyncType.allCases)
        SyncType.allCases.forEach { syncType in
            syncStatuses[syncType]!.lastSyncAttempted = newDate
        }
        
        let (lastModifiedResponse, status, _) = HttpRequester.getSync("sync/last-modified", LastModifiedTimes.self)
        if status < 200 || status >= 300 {
            GGSyncLog.error("Could not get last modified times from the API. Not syncing")
            syncRunning = false
            return false
        }
        
        guard let lastModifiedTimestamps = lastModifiedResponse?.lastModifiedTimestamps else {
            GGSyncLog.error("Unable to parse last modified timestamps from the API. Not syncing")
            syncRunning = false
            return false
        }
        
        var semaphores: [DispatchSemaphore] = []
        
        SyncType.allCases.forEach { syncType in
            var syncStatus = syncStatuses[syncType]!
            let lastSynced = syncStatus.lastSynced
            let lastModified = lastModifiedTimestamps[syncType.rawValue]!

            if lastSynced < lastModified {
                GGSyncLog.debug("Syncing \(syncType). Last synced on device: \(lastSynced) and last modified on the server: \(lastModified)")
                
                let semaphore = DispatchSemaphore(value: 0)
                semaphores.append(semaphore)
                
                DispatchQueue.global().async {
                    var syncSuccess = true
                    
                    switch syncType {
                    case .track:
                        syncSuccess = TrackSynchronizer.sync(syncStatus, maximum, pageCompleteCallback)
                    case .playlist:
                        syncSuccess = PlaylistSynchronizer.sync(syncStatus, maximum, pageCompleteCallback)
                    case .playlistTrack:
                        syncSuccess = PlaylistTrackSynchronizer.sync(syncStatus, maximum, pageCompleteCallback)
                    case .user:
                        syncSuccess = UserSynchronizer.sync(syncStatus, maximum, pageCompleteCallback)
                    case .reviewSource:
                        syncSuccess = ReviewSourceSynchronizer.sync(syncStatus, maximum, pageCompleteCallback)
                    }
                    
                    if !syncSuccess {
                        GGSyncLog.error("Unable to completely sync entity \(syncStatus.syncType). Not saving sync date")
                    } else {
                        syncStatus.lastSynced = newDate
                        SyncStatusDao.save(syncStatus)
                    }
                    
                    semaphore.signal()
                }
            } else {
                SyncStatusDao.save(syncStatus)
            }
        }
        
        semaphores.forEach { $0.wait() }
        
        if offlineModeEnabled {
            GGSyncLog.warning("Offline mode was enabled while we were syncing.")
        }
        
        // Even if the sync is not entirely successful, we still want to update the last sync time so we dont just spam if there's an error
        lastSync = newDate
        syncRunning = false

        GGSyncLog.info("Sync finished")
        
        DispatchQueue.global().async {
            TrackService.retryFailedListens()
        }
        
        return true
    }
}

struct EntityChangeResponse<T: Codable>: Codable {
    let content: EntityChangeContent<T>
    let pageable: EntitySyncPagination
}

struct EntityChangeContent<T: Codable>: Codable {
    let new: Array<T>
    let modified: Array<T>
    let removed: Array<Int>
}

struct EntitySyncPagination: Codable {
    let offset: Int
    let pageSize: Int
    let pageNumber: Int
    let totalPages: Int
    let totalElements: Int
}

struct LastModifiedTimes: Codable {
    // Can't deserialize to a raw value apparently, and I don't feel like renaming these enums to match the server since it feels wrong.
    // So I'm using a String here as it only needs to be converted to a SyncType in one place anyway...
//    let lastModifiedTimestamps: [SyncType: Date]
    let lastModifiedTimestamps: [String: Date]
}

File: android/app/src/main/java/com/gorilla/gorillagroove/service/sync/ServerSynchronizer.kt
package com.gorilla.gorillagroove.service.sync

import com.gorilla.gorillagroove.database.GorillaDatabase
import com.gorilla.gorillagroove.database.entity.DbSyncStatus
import com.gorilla.gorillagroove.di.Network
import com.gorilla.gorillagroove.service.GGLog.logDebug
import com.gorilla.gorillagroove.service.GGLog.logError
import com.gorilla.gorillagroove.service.GGLog.logInfo
import com.gorilla.gorillagroove.ui.OfflineModeService
import com.gorilla.gorillagroove.service.GGSettings
import com.gorilla.gorillagroove.util.toMutableMap
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.time.Instant

private const val MIN_TIME_BETWEEN_SYNCS_SECONDS = 600L // 10 minutes

object ServerSynchronizer {
    private var syncRunning = false
    private var lastSync = Instant.MIN

    private val syncStatusDao get() = GorillaDatabase.syncStatusDao

    suspend fun syncWithServer(
        syncTypes: Set<SyncType> = SyncType.values().toSet(),
        abortIfRecentlySynced: Boolean = false,
        onPageSyncedHandler: SyncHandler? = null,
    ) = withContext(Dispatchers.IO) {
        if (GGSettings.offlineModeEnabled) {
            return@withContext
        }

        logInfo("Running sync with API for the types: $syncTypes")
        if (abortIfRecentlySynced && lastSync.plusSeconds(MIN_TIME_BETWEEN_SYNCS_SECONDS).isAfter(Instant.now())) {
            logDebug("Last sync was too recent. Not syncing")
            return@withContext
        }

        synchronized(this) {
            if (syncRunning) {
                return@withContext
            }

            syncRunning = true
        }

        val syncTime = Instant.now()

        val syncStatuses = syncStatusDao.findAll().map { it.syncType to it }.toMutableMap()

        syncTypes.forEach { syncType ->
            // If this is the first time we have synced, a sync status might not yet exist
            val dbSyncStatus = syncStatuses[syncType] ?: DbSyncStatus(syncType = syncType, lastSynced = null, lastSyncAttempted = null).also {
                syncStatusDao.save(it)
                syncStatuses[syncType] = it
            }

            dbSyncStatus.lastSyncAttempted = syncTime
        }

        val lastModifiedTimestamps = try {
            Network.api.getLastModifiedTimestamps().lastModifiedTimestamps
        } catch (e: Throwable) {
            logError("Failed to get last modified timestamps!", e)
            synchronized(this) {
                syncStatusDao.save(syncStatuses.values.toList())
                syncRunning = false
                return@withContext
            }
        }

        syncTypes.forEach { syncType ->
            val syncStatus = syncStatuses.getValue(syncType)
            val lastSynced = syncStatus.lastSynced ?: Instant.MIN
            val lastModified = lastModifiedTimestamps[syncType.name]

            if (lastSynced.isAfter(lastModified)) {
                syncStatusDao.save(syncStatus)
            } else {
                logInfo("About to sync type: $syncType")

                val success = when (syncType) {
                    SyncType.TRACK -> TrackSynchronizer.sync(syncStatus, syncTime, onPageSyncedHandler)
                    SyncType.PLAYLIST_TRACK -> PlaylistTrackSynchronizer.sync(syncStatus, syncTime, onPageSyncedHandler)
                    SyncType.PLAYLIST -> PlaylistSynchronizer.sync(syncStatus, syncTime, onPageSyncedHandler)
                    SyncType.USER -> UserSynchronizer.sync(syncStatus, syncTime, onPageSyncedHandler)
                    SyncType.REVIEW_SOURCE -> ReviewSourceSynchronizer.sync(syncStatus, syncTime, onPageSyncedHandler)
                }

                if (success) {
                    syncStatus.lastSynced = syncTime
                    syncStatusDao.save(syncStatus)
                } else {
                    logError("Unable to completely sync type $syncType. Not saving sync date")
                }
            }
        }

        // If we attempted to sync everything, update our last in-memory sync time even if there were failures. We don't want to just spam, especially since it probably failed from bad internet.
        if (syncTypes.size == SyncType.values().size) {
            lastSync = syncTime
        }

        synchronized(this) { syncRunning = false }

        logInfo("Sync has finished running")

        OfflineModeService.downloadAlwaysOfflineTracks()
    }
}

enum class SyncType {
    TRACK,
    PLAYLIST_TRACK,
    PLAYLIST,
    USER,
    REVIEW_SOURCE
    ;
}

data class LastModifiedTimesResponse(
    val lastModifiedTimestamps: Map<String, Instant>
)

data class EntityChangeResponse<T>(
    val content: EntityChangeContent<T>,
    val pageable: EntityPagination,
) {
    data class EntityChangeContent<T>(
        val new: List<T>,
        val modified: List<T>,
        val removed: List<Long>,
    )
}

data class EntityPagination(
    val offset: Int,
    val pageSize: Int,
    val pageNumber: Int,
    val totalPages: Int,
    val totalElements: Int,
)
